# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/webbrowserpersist/nsWebBrowserPersist.h
# Commit: 1b44712c42dc
# Full Hash: 1b44712c42dc0b1504e4b51f06ef944b611654ea
# Author: Gijs Kruitbosch <gijskruitbosch@gmail.com>
# Date: 2020-09-03 09:45:53
# Regressor Bug: 1658202
# File Overlap Count: 2
# Description:
#   Bug 1658202 - move writes from onDataAvailable away from the main thread, r=valentin
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D88730
# ==============================================================================

diff -r 033e491241b1 -r 1b44712c42dc dom/webbrowserpersist/nsWebBrowserPersist.h
--- a/dom/webbrowserpersist/nsWebBrowserPersist.h	Wed Sep 02 19:45:37 2020 +0000
+++ b/dom/webbrowserpersist/nsWebBrowserPersist.h	Wed Sep 02 23:15:06 2020 +0000
@@ -18,6 +18,7 @@
 #include "nsIChannel.h"
 #include "nsIProgressEventSink.h"
 #include "nsIFile.h"
+#include "nsIThreadRetargetableStreamListener.h"
 #include "nsIWebProgressListener2.h"
 #include "nsIWebBrowserPersist.h"
 #include "nsIWebBrowserPersistDocument.h"
@@ -33,6 +34,7 @@
 class nsWebBrowserPersist final : public nsIInterfaceRequestor,
                                   public nsIWebBrowserPersist,
                                   public nsIStreamListener,
+                                  public nsIThreadRetargetableStreamListener,
                                   public nsIProgressEventSink,
                                   public nsSupportsWeakReference {
   friend class nsEncoderNodeFixup;
@@ -41,12 +43,13 @@
  public:
   nsWebBrowserPersist();
 
-  NS_DECL_ISUPPORTS
+  NS_DECL_THREADSAFE_ISUPPORTS
   NS_DECL_NSIINTERFACEREQUESTOR
   NS_DECL_NSICANCELABLE
   NS_DECL_NSIWEBBROWSERPERSIST
   NS_DECL_NSIREQUESTOBSERVER
   NS_DECL_NSISTREAMLISTENER
+  NS_DECL_NSITHREADRETARGETABLESTREAMLISTENER
   NS_DECL_NSIPROGRESSEVENTSINK
 
   // Private members
@@ -146,8 +149,10 @@
    */
   nsCOMPtr<nsIWebProgressListener2> mProgressListener2;
   nsCOMPtr<nsIProgressEventSink> mEventSink;
+  mozilla::Mutex mOutputMapMutex;
   nsClassHashtable<nsISupportsHashKey, OutputData> mOutputMap;
   nsClassHashtable<nsISupportsHashKey, UploadData> mUploadList;
+  nsCOMPtr<nsISerialEventTarget> mBackgroundQueue;
   nsClassHashtable<nsCStringHashKey, URIData> mURIMap;
   nsCOMPtr<nsIWebBrowserPersistURIMap> mFlatURIMap;
   nsTArray<mozilla::UniquePtr<WalkData>> mWalkStack;
@@ -156,13 +161,17 @@
   nsTArray<nsCString> mFilenameList;
   bool mFirstAndOnlyUse;
   bool mSavingDocument;
-  bool mCancel;
+  // mCancel is used from both the main thread, and (inside OnDataAvailable)
+  // from a background thread.
+  mozilla::Atomic<bool> mCancel;
   bool mCompleted;
   bool mStartSaving;
   bool mReplaceExisting;
   bool mSerializingOutput;
   bool mIsPrivate;
-  uint32_t mPersistFlags;
+  // mPersistFlags can be modified on the main thread, and can be read from
+  // a background thread when OnDataAvailable calls MakeOutputStreamFromFile.
+  mozilla::Atomic<uint32_t> mPersistFlags;
   nsresult mPersistResult;
   int64_t mTotalCurrentProgress;
   int64_t mTotalMaxProgress;
